File: tests/expected/infer-ops.jl

function foo()
    a = 10
    b = 20
    c1 = (a + b)
    c2 = (a - b)
    c3 = (a * b)
    c4 = (a / b)
    c5 = -(a)
    d = 2.0
    e1 = (a + d)
    e2 = (a - d)
    e3 = (a * d)
    e4 = (a / d)
    f = -3.0
    g = -(a)
end

function add1(x::c_int8, y::c_int8)::c_int16
    return (x + y)
end

function add2(x::c_int16, y::c_int16)::c_int32
    return (x + y)
end

function add3(x::c_int32, y::c_int32)::c_int64
    return (x + y)
end

function add4(x::c_int64, y::c_int64)::c_int64
    return (x + y)
end

function add5(x::c_uint8, y::c_uint8)::c_uint16
    return (x + y)
end

function add6(x::c_uint16, y::c_uint16)::c_uint32
    return (x + y)
end

function add7(x::c_uint32, y::c_uint32)::c_uint64
    return (x + y)
end

function add8(x::c_uint64, y::c_uint64)::c_uint64
    return (x + y)
end

function add9(x::c_int8, y::c_uint16)::c_uint32
    return (x + y)
end

function sub(x::c_int8, y::c_int8)::c_int8
    return (x - y)
end

function mul(x::c_int8, y::c_int8)::c_int16
    return (x * y)
end

function fadd(x::c_int8, y::Float64)::Float64
    return (x + y)
end

File: tests/expected/infer_ops.cpp
#include <cassert>   // NOLINT(build/include_order)
#include <cstdint>   // NOLINT(build/include_order)
#include <iostream>  // NOLINT(build/include_order)

inline void foo() {
  int a = 10;
  int b = 20;
  int _c1 = a + b;
  int _c2 = a - b;
  int _c3 = a * b;
  double _c4 = a / b;
  int _c5 = -(a);
  double d = 2.0;
  double _e1 = a + d;
  double _e2 = a - d;
  double _e3 = a * d;
  double _e4 = a / d;
  double _f = -3.0;
  int _g = -(a);
}

inline int16_t add1(int8_t x, int8_t y) { return x + y; }

inline int32_t add2(int16_t x, int16_t y) { return x + y; }

inline int64_t add3(int32_t x, int32_t y) { return x + y; }

inline int64_t add4(int64_t x, int64_t y) { return x + y; }

inline uint16_t add5(uint8_t x, uint8_t y) { return x + y; }

inline uint32_t add6(uint16_t x, uint16_t y) { return x + y; }

inline uint64_t add7(uint32_t x, uint32_t y) { return x + y; }

inline uint64_t add8(uint64_t x, uint64_t y) { return x + y; }

inline uint32_t add9(int8_t x, uint16_t y) { return x + y; }

inline int8_t sub(int8_t x, int8_t y) { return x - y; }

inline int16_t mul(int8_t x, int8_t y) { return x * y; }

inline double fadd1(int8_t x, double y) { return x + y; }

inline void show() {
  assert(fadd1(6, 6.0) == 12);
  std::cout << std::string{"OK"};
  std::cout << std::endl;
}

int main(int argc, char** argv) {
  foo();
  show();
}
